radixdlt-java/src/main/java/com/radixdlt/client/core/network/RadixNetwork.java

```diff
@@ -25,28 +24,41 @@
 	/**
 	 * Hot observable which updates subscribers of new connection events
 	 */
-	private final ConnectableObservable<SimpleImmutableEntry<String, RadixClientStatus>> statusUpdates;
-
+	private final ConnectableObservable<SimpleImmutableEntry<RadixPeer, RadixClientStatus>> statusUpdates;
+	private final Observable<RadixNetworkState> networkState;
 
 	public RadixNetwork(PeerDiscovery peerDiscovery) {
 		Objects.requireNonNull(peerDiscovery);
 
 		this.peers = peerDiscovery.findPeers()
 			.retryWhen(new IncreasingRetryTimer(WebSocketException.class))
-			.doOnNext(peer -> LOGGER.info("Added to peer list: " + peer.getLocation()))
+			.doOnNext(peer -> {
+				peer.getRadixClient().fetchClientInformation();
+				LOGGER.info(String.format("Added to peer list: %s", peer.getLocation()));
+			})
 			.replay().autoConnect(2);
 
-		this.statusUpdates = peers.map(RadixPeer::getRadixClient)
+		this.statusUpdates = peers
 			.flatMap(
-				client -> client.getStatus().map(
-					status -> new SimpleImmutableEntry<>(client.getLocation(), status)
+				peer -> peer.getRadixClient().status()
+						.map(status -> new SimpleImmutableEntry<>(peer, status)
 				)
 			)
 			.publish();
 		this.statusUpdates.connect();
+
+		this.networkState = peers.flatMap(peer -> peer.status()
+				.doOnNext(status -> LOGGER.info(String.format("Peer status changed: %s", status)))
+				.map(status -> new SimpleImmutableEntry<>(peer, status)))
+				.scan(new RadixNetworkState(Collections.emptyMap()), (previousState, update) -> {
+			LinkedHashMap<RadixPeer, RadixPeerState> currentPeers = new LinkedHashMap<>(previousState.getPeers());
+			currentPeers.put(update.getKey(), update.getValue());
+
+			return new RadixNetworkState(currentPeers);
+		});
 	}
 
-	public Observable<SimpleImmutableEntry<String, RadixClientStatus>> connectAndGetStatusUpdates() {
+	public Observable<SimpleImmutableEntry<RadixPeer, RadixClientStatus>> connectAndGetStatusUpdates() {
```

Comment:
Same issue here, weird to have RadixPeer as key when RadixPeer is mutable and has state.

Reply:
After further consideration I've changed it to just return Observable<RadixPeerState>, that seems cleaner. There is technically a similar issue in RadixNetworkState where RadixPeer is used as a key, so I've changed that to an identity map to be safe (as we do want an instance of the actual RadixPeer there, otherwise it's kind of useless - could change so that you can look up peers in network via their peer state, but for now we're good I think).